components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.common.model.User;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.OAuth2TokenValidationService;
+import org.wso2.carbon.identity.oauth2.ResponseHeader;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2ClientApplicationDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationRequestDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationResponseDTO;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.identity.user.account.association.UserAccountConnectorImpl;
+import org.wso2.carbon.identity.user.account.association.dto.UserAccountAssociationDTO;
+import org.wso2.carbon.identity.user.account.association.exception.UserAccountAssociationException;
+
+import java.util.Arrays;
+import java.util.HashSet;
+
+public class AccountSwitchGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    private static final Log log = LogFactory.getLog(AccountSwitchGrantHandler.class);
+
+    public static final String TOKEN_PARAM = "token";
+    public static final String USERNAME_PARAM = "username";
+    public static final String USERSTORE_DOMAIN_PARAM = "userstore-domain";
+    public static final String TENANT_DOMAIN_PARAM = "tenant-domain";
+    private final String OAUTH_HEADER = "Bearer";
+    private final String ACCOUNT_SWITCH_TOKEN_VALIDATION_CONTEXT = "accountSwitchTokenValidationContext";
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        super.validateGrant(tokReqMsgCtx);
+
+        String token = extractParameter(TOKEN_PARAM, tokReqMsgCtx);;
+        String username = extractParameter(USERNAME_PARAM, tokReqMsgCtx);;
+        String userstoreDomain = extractParameter(USERSTORE_DOMAIN_PARAM, tokReqMsgCtx);;
+        String tenantDomain = extractParameter(TENANT_DOMAIN_PARAM, tokReqMsgCtx);;
+
+        if (StringUtils.isEmpty(token) || StringUtils.isEmpty(username) || StringUtils.isEmpty(username) ||
+                StringUtils.isEmpty(username)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Grant validation failed. Missing required parameters.");
+            }
+            ResponseHeader responseHeader = new ResponseHeader();
+            responseHeader.setKey("error-description");
+            responseHeader.setValue("Missing required parameters");
+            tokReqMsgCtx.addProperty("RESPONSE_HEADERS", new ResponseHeader[]{responseHeader});
+
+            return false;
+        }
+
+        OAuth2TokenValidationResponseDTO validationResponseDTO = validateToken(token);
```

Comment:
Verify whether its need additional validation with the new "token bindings"
@thanujalk